orderbook/src/database.rs

```diff
@@ -11,16 +15,148 @@ pub struct Trade {
     pub fee_amount: U256,
 }
 
-/// Abstraction over the database we use to store persistent data like orders and trade events.
-#[async_trait::async_trait]
-pub trait Database {
-    async fn block_number_of_most_recent_trade(&self) -> Result<u64>;
+// The pool uses an Arc internally.
+#[derive(Clone)]
+pub struct Database {
+    pool: PgPool,
+}
+
+impl Database {
+    pub fn new(uri: &str) -> Result<Self> {
+        Ok(Self {
+            pool: PgPool::connect_lazy(uri)?,
+        })
+    }
+
+    pub async fn block_number_of_most_recent_trade(&self) -> Result<Option<u64>> {
+        const QUERY: &str = "SELECT MAX(block_number) FROM trades;";
+        let block_number: Option<i64> = sqlx::query_scalar(QUERY)
+            .fetch_one(&self.pool)
+            .await
+            .context("block_number_of_most_recent_trade failed")?;
+        block_number
+            .map(|block_number| block_number.try_into().context("block number is negative"))
+            .transpose()
+    }
+
     // All insertions happen in one transaction.
-    async fn insert_trades(&self, trades: &[Trade]) -> Result<()>;
+    pub async fn insert_trades(&self, trades: Vec<Trade>) -> Result<()> {
+        let mut connection = self.pool.acquire().await?;
+        connection
+            .transaction(move |transaction| {
+                async move {
+                    insert_trades(transaction, trades.as_slice())
+                        .await
+                        .context("insert_trades failed")
+                }
+                .boxed()
+            })
+            .await?;
+        Ok(())
+    }
+
     // The deletion and all insertions happen in one transaction.
-    async fn replace_trades(
+    pub async fn replace_trades(
         &self,
         delete_from_block_number: u64,
-        new_trades: &[Trade],
-    ) -> Result<()>;
+        trades: Vec<Trade>,
+    ) -> Result<()> {
+        let mut connection = self.pool.acquire().await?;
+        connection
+            .transaction(move |transaction| {
```

Comment:
The way sqlx specifies the lifetimes for the transaction closure means we have to take ownership over the trades.